src/status_emitter/libtest_json.rs

```diff
@@ -0,0 +1,227 @@
+
+    // Adapted from test::formatters::json::write_test_start().
+    println!(r#"{{ "type": "test", "event": "started", "name": "{triaged_name}" }}\n"#);
+}
+
+// Adapted from test::formatters::json.
+/// A formatting utility used to print strings with characters in need of escaping.
+/// Base code taken form `libserialize::json::escape_str`.
+struct EscapedString<S: AsRef<str>>(S);
+
+impl<S: AsRef<str>> Display for EscapedString<S> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> ::std::fmt::Result {
+        let mut start = 0;
+
+        for (i, byte) in self.0.as_ref().bytes().enumerate() {
+            let escaped = match byte {
+                b'"' => "\\\"",
+                b'\\' => "\\\\",
+                b'\x00' => "\\u0000",
+                b'\x01' => "\\u0001",
+                b'\x02' => "\\u0002",
+                b'\x03' => "\\u0003",
+                b'\x04' => "\\u0004",
+                b'\x05' => "\\u0005",
+                b'\x06' => "\\u0006",
+                b'\x07' => "\\u0007",
+                b'\x08' => "\\b",
+                b'\t' => "\\t",
+                b'\n' => "\\n",
+                b'\x0b' => "\\u000b",
+                b'\x0c' => "\\f",
+                b'\r' => "\\r",
+                b'\x0e' => "\\u000e",
+                b'\x0f' => "\\u000f",
+                b'\x10' => "\\u0010",
+                b'\x11' => "\\u0011",
+                b'\x12' => "\\u0012",
+                b'\x13' => "\\u0013",
+                b'\x14' => "\\u0014",
+                b'\x15' => "\\u0015",
+                b'\x16' => "\\u0016",
+                b'\x17' => "\\u0017",
+                b'\x18' => "\\u0018",
+                b'\x19' => "\\u0019",
+                b'\x1a' => "\\u001a",
+                b'\x1b' => "\\u001b",
+                b'\x1c' => "\\u001c",
+                b'\x1d' => "\\u001d",
+                b'\x1e' => "\\u001e",
+                b'\x1f' => "\\u001f",
+                b'\x7f' => "\\u007f",
+                _ => {
+                    continue;
+                }
+            };
+
+            if start < i {
+                f.write_str(&self.0.as_ref()[start..i])?;
+            }
+            f.write_str(escaped)?;
+
+            start = i + 1;
+        }
+
+        if start != self.0.as_ref().len() {
+            f.write_str(&self.0.as_ref()[start..])?;
+        }
+        Ok(())
+    }
+}
+
+// MAINTENANCE REGION END
+
+/// A libtest JSON output emitter.
+#[derive(Clone)]
+pub struct LibtestJSON;
+
+impl StatusEmitter for LibtestJSON {
+    /// Create a report about the entire test run at the end.
+    fn finalize(
+        &self,
+        failed: usize,
+        succeeded: usize,
+        ignored: usize,
+        filtered: usize,
+        aborted: bool,
+    ) -> Box<dyn Summary> {
+        let status = if aborted || failed > 0 {
+            String::from("failed")
+        } else {
+            String::from("ok")
+        };
+
+        emit_suite_start(failed + ignored + succeeded);
```

Comment:
😆 shouldn't this be emitted first like other tests do it? Which is like the whole problem with counting tests

Reply:
Either do not emit a count, or add logic to ui_test to do a dry run that just collects the number of tests

Reply:
It turns out that as long as the event is emitted at some point, this will satisfy our use case.

If we want to be chronologically correct, then indeed, a "suite start" has to be the first event emitted. I am not sure whether adding dry run logic is worth the complexity and maintenance burden. (The number of tests is important!)

Reply:
> (The number of tests is important!)

I really question this factoid 😆 All you are doing is sanity checking the test suite on whether it correctly precomputed the number of tests. You even completely subverted this sanity check here, as you are not emitting the number of tests after running the tests. So I would recommend to just stop relying on this number for the suite start and *only* checking the number at the suite end (and checking that against the individual test jsons you got)

Reply:
This will break the test suite renderer in https://github.com/rust-lang/rust/blob/8f274491a172e635abc447eaa44438e9239b0f05/src/bootstrap/src/utils/render_tests.rs#L310 and erase all recorded tests before printing that none have been executed

Out of different reasons I started messing with the compiletest suite in rustc, and it turns out it does a lot of work ahead of time just to print a reliable number (like enough work to just run 50 tests completely from start to finish in the same time). And I found a bunch of comments hinting at things being this way just for legacy reasons. So I started a discussion in https://rust-lang.zulipchat.com/#narrow/channel/122651-general/topic/compiletest.20tuning/with/515040266, which may result in us changing what compiletest generally outputs

Reply:
Done.

Reply:
Oh we can still emit it, just without the number of tests field

Reply:
OK, will do!